falcon/util/mediatypes.py

```diff
@@ -88,3 +92,221 @@ def parse_header(line: str) -> Tuple[str, Dict[str, str]]:
+    ) -> None:
+        self.main_type = main_type
+        self.subtype = subtype
+        self.quality = quality
+        self.params = params
+
+    @classmethod
+    def parse(cls, media_range: str) -> _MediaRange:
+        try:
+            main_type, subtype, params = _parse_media_type_header(media_range)
+        except errors.InvalidMediaType as ex:
+            raise errors.InvalidMediaRange(
+                'The media range value must contain type/subtype.'
+            ) from ex
+
+        # NOTE(vytas): We don't need to special-case Q since the above
+        #   parse_header always lowercases parameters.
+        q = params.pop('q', 1.0)
+
+        try:
+            quality = float(q)
+        except (TypeError, ValueError) as ex:
+            # NOTE(vytas): RFC 9110, Section 12.4.2:
+            #   weight = OWS ";" OWS "q=" qvalue
+            #   qvalue = ( "0" [ "." 0*3DIGIT ] ) / ( "1" [ "." 0*3("0") ] )
+            raise errors.InvalidMediaRange(cls._Q_VALUE_ERROR_MESSAGE) from ex
+
+        if not (0.0 <= quality <= 1.0) or not math.isfinite(quality):
+            raise errors.InvalidMediaRange(cls._Q_VALUE_ERROR_MESSAGE)
+
+        # NOTE(vytas): RFC 9110, Section 12.4.2 states that a sender of qvalue
+        #   MUST NOT generate more than three digits after the decimal point,
+        #   but we are more permissive here, and opt not to spend any extra CPU
+        #   cycles, if we have already managed to convert the value to float.
+
+        return cls(main_type, subtype, quality, params)
+
+    def match_score(self, media_type: _MediaType) -> Tuple[int, int, int, int, float]:
+        if self.main_type == '*' or media_type.main_type == '*':
+            main_matches = 0
+        elif self.main_type != media_type.main_type:
+            return self._NOT_MATCHING
+        else:
+            main_matches = 1
+
+        if self.subtype == '*' or media_type.subtype == '*':
+            sub_matches = 0
+        elif self.subtype != media_type.subtype:
+            return self._NOT_MATCHING
+        else:
+            sub_matches = 1
+
+        mr_pnames = frozenset(self.params)
+        mt_pnames = frozenset(media_type.params)
+
+        exact_match = 0 if mr_pnames.symmetric_difference(mt_pnames) else 1
+
+        matching = mr_pnames & mt_pnames
+        for pname in matching:
+            if self.params[pname] != media_type.params[pname]:
+                return self._NOT_MATCHING
+
+        return (main_matches, sub_matches, exact_match, len(matching), self.quality)
+
+    def __repr__(self) -> str:
+        q = self.quality
+        return f'MediaRange<{self.main_type}/{self.subtype}; {q=}; {self.params}>'
+
+
+# PERF(vytas): It is possible to cache a classmethod too, but the invocation is
+#   less efficient, especially in the case of a cache hit.
+_parse_media_type = functools.lru_cache(_MediaType.parse)
+_parse_media_range = functools.lru_cache(_MediaRange.parse)
```

Comment:
Not sure I'm following this one. Why would that be a problem?
Or do you mean that we couldn't put the cache decorator directly on the `@classmethod`s in question?

In any case, as long as we keep these cached utilities private, we could make the classes public?

Reply:
That I meant is the if return values to the users that come from these method, then they should no longer be cached, since the user may modify the cached instance

Reply:
If the methods are still private, the users should not be accessing them, so they wouldn't get the same instance.

I mean
```python
>>> class SomeClass:
...     def __init__(self, value):
...         self.value = value
...     @classmethod
...     def parse(cls, value):
...         return cls(str(value))
...         
>>> import functools
>>> _cached_parse = functools.lru_cache(SomeClass.parse)
>>> 
>>> sc1 = _cached_parse(1)
>>> sc2 = _cached_parse(2)
>>> sc1.value
'1'
>>> sc2.value
'2'
>>> sc_u1 = SomeClass.parse(1)
>>> sc_u1.value = 'modified'
>>> sc1.value
'1'
>>> sc1 is sc_u1
False
```

Reply:
I added a note that we need to keep these private cached methods, or if we move caching directly to `@classmethod`s, we need to use frozen dataclasses.

Reply:
Fronzen dataclasses are not enough since params is a dict.